### ISSUE TITLE
cannot deploy to heroku

### ISSUE THREAD

**Issue by awoodrum87** [ga-wdi-boston/full-stack-project#786]
I am able to `git push heroku master` but run into an error when I run `heroku run rake db:migrate`

The error is:
```
rake aborted!
ActiveRecord::DuplicateMigrationNameError: 

Multiple migrations have the name CreateReviewers.

/app/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1236:in `validate'
/app/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1107:in `initialize'
/app/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1006:in `new'
/app/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1006:in `up'
/app/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:984:in `migrate'
/app/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/tasks/database_tasks.rb:161:in `migrate'
/app/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/railties/databases.rake:58:in `block (2 levels) in <top (required)>'
/app/vendor/bundle/ruby/2.3.0/gems/rake-12.0.0/exe/rake:27:in `<top (required)>'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/cli/exec.rb:74:in `load'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/cli/exec.rb:74:in `kernel_load'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/cli/exec.rb:27:in `run'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/cli.rb:332:in `exec'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/vendor/thor/lib/thor.rb:359:in `dispatch'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/cli.rb:20:in `dispatch'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/vendor/thor/lib/thor/base.rb:440:in `start'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/cli.rb:11:in `start'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/exe/bundle:34:in `block in <top (required)>'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/friendly_errors.rb:100:in `with_friendly_errors'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/exe/bundle:26:in `<top (required)>'
/app/bin/bundle:3:in `load'
/app/bin/bundle:3:in `<main>'
Tasks: TOP => db:migrate
```

I'm unsure how to resolve this because I certainly don't want to delete a migration..

**Comment by payne-chris-r** [comment]
Did you edit CreateReviewers at any point?

**Comment by BenGitsCode** [comment]
Do you see two `CreateReviewers` migrations in your editor?

**Comment by tvcrawley** [comment]
Looks like an additional `CreateReviewers` was added in during the merge from feature to master. 

What do you think you could do to fix that?

**Comment by awoodrum87** [comment]
Ima going to look into these and report back.

**Comment by BenGitsCode** [comment]
Awesome! Don't hesitate to update it whenever something changes.

**Comment by awoodrum87** [comment]
OK! I was confused because I could only see 1 `CreateReviewers` migration file. Then I realize, I'm on my feature branch and my master is what get pushed! Aha! So to my master branch I go. I have the 2 `CreateReviewers` migrations. I delete the one I do not want and attempt to push. The console said up to date. 

So I go to my Github repo and see that there are 2  `CreateReviewers` migration files and said, 'eh what the heck, I'll just delete it directly' and don't you know..I deleted the wrong migration file. 👎 

I cannot, despite all my Googling prowess, figure out how to recover this. Is there a way you know of to do this?

I owe all of you drinks. And probably money.

**Comment by danman01** [comment]
You deleted the wrong migration file, because there is a new error?

Do you have both migration locally in your master branch, thus allowing you to recover it from your local version? Also, your commit history should have the file if you roll back to a previous commit on github (I think...)

**Comment by awoodrum87** [comment]
@danman01 I have 2 reviewERS migration and a reviews migration. I meant to delete on of the reviewers but accidentally deleted the reviews migration. My local is good. My github is not

**Comment by jordanallain** [comment]
did you delete it on github or on heroku?

**Comment by awoodrum87** [comment]
I deleted reviews (which I did not want to delete) on Github.

**Comment by danman01** [comment]
So you can push your local master to github, and that should restore the file on github. However it may say you have changes upstream and ask you to first pull, then push. Instead of doing that, you should do a force push to overwrite what you have on github with what you have on local, but [mandatory force push warning] only if you are absolutely sure you want to overwrite what you have on github with local. Use `--force-with-lease` instead of `--force` as it's a little safer.

**Comment by awoodrum87** [comment]
thanks danny, let me try that and get back to you

**Comment by awoodrum87** [comment]
so I am 100% sure what is local is what I want to be on my remote. i ran`git push --force-with-lease`

and got this error. I just want to make sure I won't destroy anything if I use the suggestion. 

```
~/wdi/projects/full-stack-api (master)
$ git push --force-with-lease
fatal: The current branch master has no upstream branch.
To push the current branch and set the remote as upstream, use

    git push --set-upstream origin master

```

**Comment by danman01** [comment]
Ah, yes that should be fine. You could also do git `push origin master --force-with-lease`, but setting upstream will allow you to simply do `git push` and it'll auto push to `origin master` each time.

**Comment by awoodrum87** [comment]
```git push --set-upstream origin master```
returns this error:
```
 git push --set-upstream origin master
To https://github.com/awoodrum87/angies-other-list-api
 ! [rejected]        master -> master (fetch first)
error: failed to push some refs to 'https://github.com/awoodrum87/angies-other-list-api'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first integrate the remote changes
hint: (e.g., 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```
I obviously do not want to pull so the git suggestion will not work. 
```git push origin master --force-with-lease```
returned this:
```
$ git push origin master --force-with-lease
To https://github.com/awoodrum87/angies-other-list-api
 ! [rejected]        master -> master (stale info)
error: failed to push some refs to 'https://github.com/awoodrum87/angies-other-list-api'
```

**Comment by danman01** [comment]
Here's an article explaining how to get past this stale info warning, basically either involves using `git fetch` to trick git into thinking you have indeed updated, or using `update-ref` to do some more advanced git manipulation: https://developer.atlassian.com/blog/2015/04/force-with-lease/

However, since we know this is just one file in question, why not copy that file (`cp filename otherfilename`) and then do the git pull that it wants. Then put it back, commit and push: `mv otherfilename filename`, `git commit` `git push origin master` and I think that should solve it.

**Comment by BenGitsCode** [comment]
@danman01 would that be just as/more easily achieved by stashing the changes and reapplying them after the pull?

**Comment by danman01** [comment]
@BenGitsCode I'm not sure there are changes to be stashed, but yea, stash is good too as that allows you to keep what you've got locally, then reapply on top of the new stuff fetched from remote.

**Comment by awoodrum87** [comment]
so should i copy the file i deleted on the remote into a new file? an existing file? The extra file that i tried to delete on my remote doesn't exist locally.

**Comment by danman01** [comment]
Ah, so it doesn't exist locally or remotely. Does it exist in a previous commit? Can you run `git log` and then checkout a previous commit with `git checkout someCommitSha` and grab the file that way?

**Comment by awoodrum87** [comment]
so we can deploy to heroku, but can't actually use the back end in deployment. progress!

**Comment by awoodrum87** [comment]
As a reminder, here is the error when pushing master:
```$ git push origin master
To https://github.com/awoodrum87/angies-other-list-api
 ! [rejected]        master -> master (fetch first)
error: failed to push some refs to 'https://github.com/awoodrum87/angies-other-list-api'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first integrate the remote changes
hint: (e.g., 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
~/wdi/projects/full-stack-api (master)
```
simply following these instructions won't work because a file was deleted from my remote that is required.

**Comment by awoodrum87** [comment]
update:
-I recreated a remote repo and a heroku app 
-I reset my remotes in the command line
-I pushed my code successfully using `git push origin master` and `git push heroku master`
-I ran the heroku db:migrate successfully
-My deployed app still is not working, so I checked my dbs using `heroku pg:psql` and discovered that I have no tables. the message I get is:
```
angies-other-list::DATABASE=> \d
No relations found.
```

**Comment by BenGitsCode** [comment]
Did you run `heroku run rake db:create`

**Comment by awoodrum87** [comment]
```$ heroku run rake db:create
Running rake db:create on ⬢ angies-other-list... up, run.8709 (Free)
FATAL:  permission denied for database "postgres"
DETAIL:  User does not have CONNECT privilege.
Couldn't create database for {"adapter"=>"postgresql", "username"=>"oalylweervnymt", "password"=>"5cc51e167e469c973d8208e393eb0178a2c18d05a8f981271ed40e90f5714172", "port"=>5432, "database"=>"d2h8nvduvas727", "host"=>"ec2-23-21-235-142.compute-1.amazonaws.com"}
rake aborted!```

**Comment by BenGitsCode** [comment]
Ah. Try running `heroku pg --help` and see if that provides any helpful commands.

**Comment by awoodrum87** [comment]
so I found a credentials help category, but it just returned the username and password. But it returned the same username and password as the error above, so I guess that is good.

```$ heroku pg:credentials
Connection info string:
   "dbname=d2h8nvduvas727 host=ec2-23-21-235-142.compute-1.amazonaws.com port=5432 user=oalylweervnymt password=5cc51e167e469c973d8208e393eb0178a2c18d05a8f981271ed40e90f5714172 sslmode=require"
Connection URL:
   postgres://oalylweervnymt:5cc51e167e469c973d8208e393eb0178a2c18d05a8f981271ed40e90f5714172@ec2-23-21-235-142.compute-1.amazonaws.com:5432/d2h8nvduvas727```

**Comment by jrhorn424** [comment]
It's a bit hard to follow since there are about three different issues in this thread.

For the most recent, I believe resetting the database entirely will achieve what you want.

https://www.google.com/search?client=safari&rls=en&q=heroku+reset+database&ie=UTF-8&oe=UTF-8

**Comment by awoodrum87** [comment]
Hi @jrhorn424 , 
Thanks for the response. Here is a summary of what I did and the results. as well as the terminal history of those actions.

1. reset db - no issues
2. heroku db:migrate - aborted multiple migration file error
3. I reconcile the files
4. try again. get an error i've never seen. Stack overflow says it's a weird quirk and to try again
5. heroku db: migrate again -aborted multiple migration file error.

It seems like I just re-ignited the problem. 

```
$ heroku pg:reset DATABASE
 ▸    WARNING: Destructive action
 ▸    postgresql-objective-65619 will lose all of its data
 ▸    
 ▸    To proceed, type angies-other-list or re-run this
 ▸    command with --confirm angies-other-list

> angies-other-list
Resetting postgresql-objective-65619... done
~/wdi/projects/full-stack-api (master)
$ heroku run rake db:migrate
Running rake db:migrate on ⬢ angies-other-list... up, run.1288 (Free)
rake aborted!
ActiveRecord::DuplicateMigrationNameError: 

Multiple migrations have the name CreateReviewers.

/app/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1236:in `validate'
/app/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1107:in `initialize'
/app/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1006:in `new'
/app/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1006:in `up'
/app/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:984:in `migrate'
/app/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/tasks/database_tasks.rb:161:in `migrate'
/app/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/railties/databases.rake:58:in `block (2 levels) in <top (required)>'
/app/vendor/bundle/ruby/2.3.0/gems/rake-12.0.0/exe/rake:27:in `<top (required)>'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/cli/exec.rb:74:in `load'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/cli/exec.rb:74:in `kernel_load'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/cli/exec.rb:27:in `run'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/cli.rb:332:in `exec'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/vendor/thor/lib/thor.rb:359:in `dispatch'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/cli.rb:20:in `dispatch'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/vendor/thor/lib/thor/base.rb:440:in `start'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/cli.rb:11:in `start'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/exe/bundle:34:in `block in <top (required)>'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/friendly_errors.rb:100:in `with_friendly_errors'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/exe/bundle:26:in `<top (required)>'
/app/bin/bundle:3:in `load'
/app/bin/bundle:3:in `<main>'
Tasks: TOP => db:migrate
(See full trace by running task with --trace)
~/wdi/projects/full-stack-api (master)
$ git remote -v
heroku	https://git.heroku.com/angies-other-list.git (fetch)
heroku	https://git.heroku.com/angies-other-list.git (push)
origin	git@github.com:awoodrum87/api-angies-other-list.git (fetch)
origin	git@github.com:awoodrum87/api-angies-other-list.git (push)
~/wdi/projects/full-stack-api (master)
$ atom .
~/wdi/projects/full-stack-api (master)
$ heroku run rake db:migrate
Running rake db:migrate on ⬢ angies-other-list... !
 ▸    ENOTFOUND: getaddrinfo ENOTFOUND api.heroku.com api.heroku.com:443
~/wdi/projects/full-stack-api (master)
$ heroku run rake db:migrate
Running rake db:migrate on ⬢ angies-other-list... up, run.5794 (Free)
rake aborted!
ActiveRecord::DuplicateMigrationNameError: 

Multiple migrations have the name CreateReviewers.

/app/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1236:in `validate'
/app/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1107:in `initialize'
/app/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1006:in `new'
/app/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1006:in `up'
/app/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:984:in `migrate'
/app/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/tasks/database_tasks.rb:161:in `migrate'
/app/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/railties/databases.rake:58:in `block (2 levels) in <top (required)>'
/app/vendor/bundle/ruby/2.3.0/gems/rake-12.0.0/exe/rake:27:in `<top (required)>'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/cli/exec.rb:74:in `load'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/cli/exec.rb:74:in `kernel_load'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/cli/exec.rb:27:in `run'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/cli.rb:332:in `exec'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/vendor/thor/lib/thor.rb:359:in `dispatch'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/cli.rb:20:in `dispatch'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/vendor/thor/lib/thor/base.rb:440:in `start'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/cli.rb:11:in `start'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/exe/bundle:34:in `block in <top (required)>'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/lib/bundler/friendly_errors.rb:100:in `with_friendly_errors'
/app/vendor/bundle/ruby/2.3.0/gems/bundler-1.13.7/exe/bundle:26:in `<top (required)>'
/app/bin/bundle:3:in `load'
/app/bin/bundle:3:in `<main>'
Tasks: TOP => db:migrate
(See full trace by running task with --trace)
~/wdi/projects/full-stack-api (master)
$ git status
On branch master
Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	deleted:    db/migrate/20170427221637_create_reviewers.rb

no changes added to commit (use "git add" and/or "git commit -a")
~/wdi/projects/full-stack-api (master)
$ git rm db/migrate/20170427224121_add_reviewer_to_reviews.rb 
rm 'db/migrate/20170427224121_add_reviewer_to_reviews.rb'
~/wdi/projects/full-stack-api (master)
$ git status
On branch master
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	deleted:    db/migrate/20170427224121_add_reviewer_to_reviews.rb

Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	deleted:    db/migrate/20170427221637_create_reviewers.rb

~/wdi/projects/full-stack-api (master)
$ git commit 
[master ae7bee7] rm extra migration
 1 file changed, 5 deletions(-)
 delete mode 100644 db/migrate/20170427224121_add_reviewer_to_reviews.rb
~/wdi/projects/full-stack-api (master)
$ git push origin master
To github.com:awoodrum87/api-angies-other-list.git
 ! [rejected]        master -> master (fetch first)
error: failed to push some refs to 'git@github.com:awoodrum87/api-angies-other-list.git'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first integrate the remote changes
hint: (e.g., 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
~/wdi/projects/full-stack-api (master)
$ git pull origin master
error: Cannot pull with rebase: You have unstaged changes.
~/wdi/projects/full-stack-api (master)
$ git rebase master
Cannot rebase: You have unstaged changes.
Please commit or stash them.
~/wdi/projects/full-stack-api (master)
$ 

```

**Comment by gaand** [comment]
Did you run git status after running git commit?

**Comment by awoodrum87** [comment]
I don't think so. Here is what I just tried:
-git status (showed deleted file that needed commited)
-git rm <filename>
-git commit 
-git push (got the following error:
```
git push origin master
To github.com:awoodrum87/api-angies-other-list.git
 ! [rejected]        master -> master (fetch first)
error: failed to push some refs to 'git@github.com:awoodrum87/api-angies-other-list.git'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first integrate the remote changes
hint: (e.g., 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```
- git pull origin master
-git push origin master
-heroku db:reset 
-heroku run rake db:migrate (i get the multiple migration error)
-git push heroku master
-heroku run rake db:migrate - IT WORKS!!!
-test site and have the following error in my console when trying to login or create an account:
```index.html:1 XMLHttpRequest cannot load https://angies-other-list.herokuapp.com//sign-up/. No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'https://awoodrum87.github.io' is therefore not allowed access.```

**Comment by danman01** [comment]
Great! If this is resolved, could you close it? And could you open a new issue for the CORS error?

**Comment by awoodrum87** [comment]
I am closing this out and opening an issue that better reflects the most current issue. 
Thanks everybody for all your help!
